syntax = "proto3";
package Texus;

enum CLIENT_CMD {
	CLIENT_CMD_INVALID = 0;
	CLIENT_CMD_BEGIN = 30000;

	CLIENT_PING = 30001;

	CLIENT_LOGIN_REQ = 30004; 
	CLIENT_CREATE_REQ = 30005; 
	CLIENT_ANNOUNCE_REQ = 30006; 

	CLIENT_JOIN_ROOM_REQ = 30007;//客户端请求加入特定房间的请求
	CLIENT_QUIT_ROOM_REQ = 30008;//客户端请求退出房间的请求

	CLIENT_GAMESTART_REQ = 30009;
	
	CLIENT_CMD_END = 33000;
}

enum SERVER_CMD {
	SERVER_CMD_INVALID = 0;
	SERVER_CMD_BEGIN = 1000;

	SERVER_PONG = 1001;

	SERVER_LOGIN_RSP = 1004;
	SERVER_CREATE_RSP = 1005;
	SERVER_ANNOUNCE_RSP = 1006;

	SERVER_JUDGE_JOIN_RSP = 1007;//服务端仲裁是否允许客户加入特定房间的仲裁结果反馈
	SERVER_QUITROOM_RSP = 1008;//服务端反馈退出房间结果的告知

	SERVER_BROADCAST_SEATTABLE = 1009;

	SERVER_GAMESTART_RSP = 1010;
	//SERVER_BROADCAST_GAMESTART = 1011;

	GM_OPERATE_RSP = 1100;

	SERVER_CMD_END = 10000;
}

enum PROTO_RESULT_CODE{
	SERVER_LOGIN_RESULT_OK = 0;
	SERVER_LOGIN_RESULT_FAIL = 1;

	SERVER_LOGIN_RESULT_FAIL_NO_SUCH_PLAYER = -2;
	SERVER_LOGIN_RESULT_FAIL_ALREADY_IN_GAME = 3;
	SERVER_LOGIN_RESULT_FAIL_WRONG = 4;	

	JOINROOM_RESULT_OK = 5;
	JOINROOM_RESULT_FAIL_EXISTING_PLAYER_INROOM = 6;
	JOINROOM_RESULT_FAIL_NO_SUCH_ROOM = 7;
	JOINROOM_RESULT_FAIL_BAD_PLAYERID = 8;
	JOINROOM_RESULT_FAIL_PLAYER_AMOUNT_EXCEED = 9;
	JOINROOM_RESULT_FAIL_NO_ENOUGH_MONEY = 10;
	JOINROOM_RESULT_FAIL_ROOM_ALREADY_IN_GAME =11;

	QUITROOM_RESULT_OK = 15;
	QUITROOM_RESULT_FAIL = 16;
	QUITROOM_RESULT_FAIL_NO_SUCH_ROOM=17;
	QUITROOM_RESULT_FAIL_INVALID_ID=18;

	GAMESTART_OK = 19;
	GAMESTART_FAIL_NO_SUCH_ROOM=20;
	GAMESTART_FAIL_NOT_ROOM_OWNER=21;
	GAMESTART_FAIL_NO_ENOUGH_USER=22;
	GAMESTART_FAIL_ROOM_ALREADY_IN_GAME=23;
}
message SeatTableItem {
	string PlayerID = 1;
	int32 SeatNumber = 2;
}

message Card {
	int32 suit=1;
	int32 rank=2;
}

message PlayerTryJoin {
	//节省时间的做法，后续可迭代修正：用户输入PlayerID作为唯一标识
	string PlayerID = 1;
	int32 RoomID = 2;
}

message PlayerJoinResult {
	string PlayerID = 1;
	int32 RoomID = 2;
	int32 JoinResult = 3;
	repeated SeatTableItem SeatTable = 4;
	int32 money = 5;
}

message PlayerTryQuitRoom {
	string PlayerID = 1;
	int32 RoomID = 2;
}

message PlayerQuitRoomResult {
	string PlayerID = 1;
	int32 RoomID = 2;
	int32 QuitResult = 3;
	repeated SeatTableItem SeatTable= 4;
	int32 money = 5;
}

message BroadcastSeatTable{
	repeated SeatTableItem SeatTable= 1;
}

message GameStart{
	string PlayerID	=1;
	int32 RoomID =2 ;
}

message GameStartResult{
	string PlayerID=1;
	int32 RoomID =2;
	int32 GameStartResult=3;
	//repeated Card flop=4;
	repeated Card hole=5;
}

message Get {
	string PlayerID = 1;
	string Password = 2;
}

message PlayerSyncData {
	bytes Name = 1;
}

message PlayerCreateReq {
	string PlayerID = 1;
	string Password = 2;
	bytes Name = 3;
}

message PlayerCreateRsp {
	int32 Result = 1;
	string PlayerID = 2;
	bytes Name = 3;
	string Reason = 4;
}